src/vmsa/mod.rs

```diff
@@ -0,0 +1,496 @@
+            self.rip = 0;
+            self.rsp = 0;
+            self.rbp = 0;
+            self.rsi = 0;
+
+            self.cs.selector = 0x9100;
+            self.cs.base = 0x91000;
+        }
+    }
+
+    // Based on logic in x86_cpu_reset() (target/i386/cpu.c)
+    /// Set VMSA values to follow initialization for a VM running as a QEMU guest.
+    pub fn init_qemu(&mut self, _cpu: u64) {
+        self.ldtr.attrib = (ATTR_P_MASK | (2 << ATTR_TYPE_SHIFT)) >> ATTR_TYPE_SHIFT;
+        self.tr.attrib = (ATTR_P_MASK | (11 << ATTR_TYPE_SHIFT)) >> ATTR_TYPE_SHIFT;
+        self.cs.attrib = (ATTR_P_MASK | ATTR_S_MASK | ATTR_CS_MASK | ATTR_R_MASK | ATTR_A_MASK)
+            >> ATTR_TYPE_SHIFT;
+        self.ds.attrib = (ATTR_P_MASK | ATTR_S_MASK | ATTR_W_MASK | ATTR_A_MASK) >> ATTR_TYPE_SHIFT;
+        self.es.attrib = (ATTR_P_MASK | ATTR_S_MASK | ATTR_W_MASK | ATTR_A_MASK) >> ATTR_TYPE_SHIFT;
+        self.ss.attrib = (ATTR_P_MASK | ATTR_S_MASK | ATTR_W_MASK | ATTR_A_MASK) >> ATTR_TYPE_SHIFT;
+        self.fs.attrib = (ATTR_P_MASK | ATTR_S_MASK | ATTR_W_MASK | ATTR_A_MASK) >> ATTR_TYPE_SHIFT;
+        self.gs.attrib = (ATTR_P_MASK | ATTR_S_MASK | ATTR_W_MASK | ATTR_A_MASK) >> ATTR_TYPE_SHIFT;
+
+        self.g_pat = 0x0007040600070406;
+    }
+
+    /// Set CPU SKU values for a given VMSA.
+    pub fn cpu_sku(&mut self, mut family: u64, mut model: u64, mut stepping: u64) {
+        stepping &= 0xf;
+        model &= 0xff;
+        family &= 0xfff;
+
+        self.rdx = stepping;
+
+        if family > 0xf {
+            self.rdx |= 0xf00 | ((family - 0x0f) << 20);
+        } else {
+            self.rdx |= family << 8;
+        }
+
+        self.rdx |= ((model & 0xf) << 4) | ((model >> 4) << 16);
+    }
+
+    /// Set VMSA reset address register values.
+    pub fn reset_addr(&mut self, ra: u32) {
+        let reset_cs = ra & 0xffff0000;
+        let reset_ip = ra & 0x0000ffff;
+
+        self.rip = u64::from(reset_ip);
+        self.cs.base = u64::from(reset_cs);
+    }
+
+    /// Read binary content from a passed filename and deserialize it into a VMSA struct.
+    pub fn from_file(filename: &str) -> VmsaRWResult {
+        let data = match std::fs::read(filename) {
+            Ok(d) => d,
+            Err(e) => return VmsaRWResult::io_err(e),
+        };
+
+        let vmsa = match bincode::deserialize(&data[..]) {
+            Ok(v) => v,
+            Err(e) => return VmsaRWResult::serialization_err(e),
+        };
+
+        VmsaRWResult::read_success(vmsa)
+    }
+
+    // Save binary VMSA content to the passed filename.
+    /// Serialize a VMSA struct and write it to a passed filename.
+    pub fn to_file(&self, filename: &str) -> VmsaRWResult {
+        let vmsa_buf = match bincode::serialize(self) {
+            Ok(v) => v,
+            Err(e) => return VmsaRWResult::serialization_err(e),
+        };
+
+        const SIZE: usize = std::mem::size_of::<Vmsa>();
+
+        // Pad to 4096 bytes
+        let buf: &mut [u8] = &mut [0; SIZE];
```

Comment:
Just saw your sevctl PR and that the test suite is busted. SIZE here should be 4096, that will fix it